EATON HID: add missing usages and paths

masterwishx opened the pull request:
Adding HE/ECO mode, Bypass On/Off for Eaton USB UPS Online Models + Adding Missing usages and paths

Continue work for #2636
Fixes #2485
Ref #2495 

Added: 

```
Variables: 

battery.charger.status: discharging
device.usb.version: 01.29
input.eco.switchable: normal
outlet.1.designator: AC OUTPUT
outlet.1.protect.status: protected
input.transfer.forced: enabled
input.eco.switchable: ECO
input.bypass.switch.on: off
etc ...

R/W variables:

input.eco.switchable
input.bypass.switch.on
etc ..

Commands: 

ecomode.disable - Take the UPS out of High Efficiency (aka ECO) mode
ecomode.enable - Put UPS in High Efficiency (aka ECO) mode
essmode.disable - Take the UPS out of Energy Saver System (aka ESS) mode
essmode.enable - Put UPS in Energy Saver System (aka ESS) mode
bypass.start - Put the UPS in bypass mode
bypass.stop - Take the UPS out of bypass mode

```

arnaudquette-eaton commented on drivers/mge-hid.c:
```diff
@@ -1258,8 +1258,10 @@ static hid_info_t mge_hid2nut[] =
-	/* { "device.usb.version", ST_FLAG_STRING, 20, "UPS.System.USB.iVersion", NULL, NULL, HU_FLAG_STATIC, stringid_conversion },*/
-
+	{ "device.usb.version", ST_FLAG_STRING, 20, "UPS.System.USB.iVersion", NULL, NULL, HU_FLAG_STATIC, stringid_conversion }, /* FIXME */
+	{ "device.usb.mode", ST_FLAG_STRING, 20, "UPS.System.USB.Mode", NULL, NULL, HU_FLAG_STATIC, stringid_conversion }, /* FIXME */
```
I don't think this one is useful, and not a string (1 to set in bootloader )

masterwishx replied:
Commented

masterwishx replied:
commented

arnaudquette-eaton commented on drivers/mge-hid.c:
```diff
@@ -1258,8 +1258,10 @@ static hid_info_t mge_hid2nut[] =
-
+	{ "device.usb.version", ST_FLAG_STRING, 20, "UPS.System.USB.iVersion", NULL, NULL, HU_FLAG_STATIC, stringid_conversion }, /* FIXME */
+	{ "device.usb.mode", ST_FLAG_STRING, 20, "UPS.System.USB.Mode", NULL, NULL, HU_FLAG_STATIC, stringid_conversion }, /* FIXME */
+	{ "device.gateway.power.rate", ST_FLAG_STRING, 20, "UPS.System.Gateway.PowerRate", NULL, NULL, HU_FLAG_STATIC, stringid_conversion }, /* FIXME */
```
ditto above, don't think it's useful, and not a string (level of power provided by the UPS to the network card

masterwishx replied:
commented

arnaudquette-eaton commented on drivers/mge-hid.c:
```diff
@@ -1292,7 +1294,7 @@ static hid_info_t mge_hid2nut[] =
 	{ "battery.charger.status", 0, 0, "UPS.BatterySystem.Charger.PresentStatus.Used", NULL, "%.0f", HU_FLAG_QUICK_POLL, eaton_abm_enabled_legacy_info }, */
 	/* This data is the actual ABM status information */
-	{ "battery.charger.status", 0, 0, "UPS.BatterySystem.Charger.Mode", NULL, "%.0f", HU_FLAG_QUICK_POLL, eaton_abm_status_info },
+	{ "battery.charger.status", 0, 0, "UPS.BatterySystem.Charger.Status", NULL, "%.0f", HU_FLAG_QUICK_POLL, eaton_abm_status_info },
```
to be tested, may be avail on some units, not others. May need both Mode and Status

Unified charger info,bitmap with :
Status on bit 0-3:
 0x0: Off
 0x1: Charging
 0x2: Floating
 0x3: On, not charging
 0x4: Discharging
 0x5-0xF : Reserved
Mode on bit 4-6:
 0: CC
 1: ABM
 2-7 : Reserved
Error on bit 7:
 0: Good
 1: In error

masterwishx replied:
Also ABM table commented needs to be changes?

arnaudquette-eaton commented on drivers/mge-hid.c:
```diff
@@ -1482,8 +1484,9 @@ static hid_info_t mge_hid2nut[] =
 	{ "outlet.1.desc", ST_FLAG_RW | ST_FLAG_STRING, 20, "UPS.OutletSystem.Outlet.[2].OutletID", NULL, "PowerShare Outlet 1", HU_FLAG_ABSENT, NULL },
 	{ "outlet.1.switchable", 0, 0, "UPS.OutletSystem.Outlet.[2].PresentStatus.Switchable", NULL, "%s", HU_FLAG_STATIC, yes_no_info },
-	{ "outlet.1.status", 0, 0, "UPS.OutletSystem.Outlet.[2].PresentStatus.SwitchOn/Off", NULL, "%s", 0, on_off_info },
+	{ "outlet.1.status", 0, 0, "UPS.OutletSystem.Outlet.[1].Status", NULL, "%s", 0, on_off_info },	
```
ditto, may need both.... this one provides more than just on/off

Global status of the load segment :
0 : Not powered
1 : Not protected
2 : Protected

masterwishx replied:
so to leave { "outlet.2.status", 0, 0, "UPS.OutletSystem.Outlet.[3].PresentStatus.SwitchOn/Off", NULL, "%s", 0, on_off_info }, ?

masterwishx replied:
0,1,2 thius mean =  NULL, "%.2f", 0, NULL }, or  NULL, "%.0f", HU_FLAG_STATIC, NULL },

arnaudquette-eaton replied:
means a lookup function, something like:

static info_lkp_t eaton_enable_disable_info[] = {
	{ 0, "not powered", NULL, NULL },
	{ 1, "not protected", NULL, NULL },
	{ 2, "protected", NULL, NULL },
	{ 0, NULL, NULL, NULL }
};

arnaudquette-eaton commented on drivers/mge-hid.c:
```diff
@@ -1482,8 +1484,9 @@ static hid_info_t mge_hid2nut[] =
-	{ "outlet.1.status", 0, 0, "UPS.OutletSystem.Outlet.[2].PresentStatus.SwitchOn/Off", NULL, "%s", 0, on_off_info },
+	{ "outlet.1.status", 0, 0, "UPS.OutletSystem.Outlet.[1].Status", NULL, "%s", 0, on_off_info },	
 	/* FIXME: should better use UPS.OutletSystem.Outlet.[1].Status? */
+	{ "outlet.1.idesignator", 0, 0, "UPS.OutletSystem.Outlet.[1].iDesignator", NULL, "%s", 0, on_off_info }, /* needs to check  */
```
stringid_conversion (thats an *i*Value, so indexed string), not on_off ;)

masterwishx replied:
changed

arnaudquette-eaton commented on drivers/mge-hid.c:
```diff
@@ -1494,8 +1497,8 @@ static hid_info_t mge_hid2nut[] =
-	/* 0: The outlet is not ECO controlled. / 1 : The outlet is ECO controlled. => Readonly! use some yes_no_info
-		{ "outlet.1.ecocontrol", 0, 0, "UPS.OutletSystem.Outlet.[2].ECOControl", NULL, "%.2f", 0, NULL },*/
+	/* 0: The outlet is not ECO controlled. / 1 : The outlet is ECO controlled. => Readonly! use some yes_no_info */
+	{ "outlet.1.ecocontrol", 0, 0, "UPS.OutletSystem.Outlet.[2].ECOControl", NULL, "%.2f", 0, NULL },
```
nut suitable name to be found!
need a lookup function, like on_off, but for:
0 : The outlet is not ECO controlled.
1 : The outlet is ECO controlled.

masterwishx replied:
added function

arnaudquette-eaton commented:
Ref: https://github.com/networkupstools/nut/issues/2495

masterwishx reviewed (commented):
> to be tested, may be avail on some units, not others. May need both Mode and Status

Unified charger info,bitmap with :
Status on bit 0-3:
 0x0: Off
 0x1: Charging
 0x2: Floating
 0x3: On, not charging
 0x4: Discharging
 0x5-0xF : Reserved
Mode on bit 4-6:
 0: CC
 1: ABM
 2-7 : Reserved
Error on bit 7:
 0: Good
 1: In error

So this is not right table?  and needs to change as you wrote above?
* Charger ABM mode                             | Charging | Floating | Resting | Discharging | Disabled   | Disabled
* UPS.BatterySystem.Charger.Mode             |     1    |    3     |   4     |      2      |     6      |    6

jimklimov commented:
Superseded by PR #2639

masterwishx commented:
@arnaudquette-eaton @jimklimov I made some changes from comments, please check when you have time

masterwishx commented on drivers/mge-hid.c:
```diff
@@ -864,17 +870,25 @@ static info_lkp_t eaton_converter_online_info[] = {
 	{ 0, NULL, NULL, NULL }
 };
 
+static info_lkp_t eaton_outlet_status_enable_disable_info[] = {
```
Added func for outlet_status

arnaudquette-eaton replied:
=> eaton_outlet_protection_status_info

arnaudquette-eaton commented on drivers/mge-hid.c:
```diff
@@ -689,6 +689,12 @@ static info_lkp_t pegasus_yes_no_info[] = {
 	{ 0, NULL, NULL, NULL }
 };
 
+static info_lkp_t eco_pegasus_yes_no_info[] = {
```
pegasus was the code name of protection station ;)

masterwishx replied:
will remove pegasus from name also from function ?

masterwishx commented on drivers/mge-hid.c:
```diff
@@ -864,17 +870,25 @@ static info_lkp_t eaton_converter_online_info[] = {
 	{ 0, NULL, NULL, NULL }
 };
 
+static info_lkp_t eaton_outlet_status_enable_disable_info[] = {
+    { 0, "not powered", NULL, NULL },
+    { 1, "not protected", NULL, NULL },
+    { 2, "protected", NULL, NULL },
+    { 0, NULL, NULL, NULL }
+};
+
```
outlet_status func added

masterwishx commented on drivers/mge-hid.c:
```diff
@@ -1438,6 +1456,9 @@ static hid_info_t mge_hid2nut[] =
 	{ "input.bypass.frequency", 0, 0, "UPS.PowerConverter.Input.[2].Frequency", NULL, "%.1f", 0, NULL },
 	{ "input.bypass.frequency.nominal", 0, 0, "UPS.Flow.[2].ConfigFrequency", NULL, "%.0f", HU_FLAG_STATIC, NULL },
 
```
@arnaudquette-eaton is it right definition , what else needed?

masterwishx commented on drivers/mge-hid.c:
```diff
@@ -1438,6 +1456,9 @@ static hid_info_t mge_hid2nut[] =
 	{ "input.bypass.frequency", 0, 0, "UPS.PowerConverter.Input.[2].Frequency", NULL, "%.1f", 0, NULL },
 	{ "input.bypass.frequency.nominal", 0, 0, "UPS.Flow.[2].ConfigFrequency", NULL, "%.0f", HU_FLAG_STATIC, NULL },
 
+	/* ECO(HE) Mode switch , maybe better = outlet.5.switchable ? */
+	{ "input.eco.switchable", ST_FLAG_RW | ST_FLAG_STRING, 8, "UPS.PowerConverter.Input.[5].Switchable", NULL, "%.0f", HU_FLAG_SEMI_STATIC, NULL },
```
@arnaudquette-eaton is it right definition , what else needed?

masterwishx replied:
ESS Maybe Converter Mode ?

arnaudquette-eaton replied:
some info on ESS: https://www.eaton.com/content/dam/eaton/products/backup-power-ups-surge-it-power-distribution/backup-power-ups/portfolio/eaton-ess-technology-brief-brochure.pdf

masterwishx replied:
Thanks a lot, do you think i can compile and test now if UPS will switch to ECO by  `upsrw` and `input.eco.switchable` to set mode 1 ?

arnaudquette-eaton replied:
up to you, you're the dev and the tester ;)

masterwishx commented:
Compiled and seems all fine ,aslo some values added like:  

```
battery.charger.status: discharging
device.usb.version: 01.29
input.eco.switchable: normal
outlet.1.designator: AC OUTPUT
outlet.1.protect.status: protected

[input.eco.switchable]
Input High Efficiency (aka ECO) mode switch
Type: STRING
Maximum length: 8
Value: normal
```

masterwishx commented:
@arnaudquette-eaton 
but seems no effect when :  `upsrw -s input.eco.switchable=1 -u user -p pass Eaton9E2000i`

masterwishx commented:
> @arnaudquette-eaton but seems no effect when : `upsrw -s input.eco.switchable=1 -u user -p pass Eaton9E2000i`

it seems other commands also not working in Unraid Terminal by upsrw however i got OK or Success after but values not changed.
Found that `override.input.transfer.low = 218` in upsd.conf and restart is working

jimklimov commented:
>  upsrw however i got OK or Success after but values not changed

Here `upsrw` only confirms that it has sent a command (no protocol/connection error along the way). You can add `-w` to wait for the result of most (all?) operations.

There was some issue logged earlier to have some sort of verification mode - to wait AND re-read to check that the setting has changed (maybe even to what was asked).

masterwishx commented:
> > upsrw however i got OK or Success after but values not changed
> 
> Here `upsrw` only confirms that it has sent a command (no protocol/connection error along the way). You can add `-w` to wait for the result of most (all?) operations.
> 
> There was some issue logged earlier to have some sort of verification mode - to wait AND re-read to check that the setting has changed (maybe even to what was asked).

I tryed both commands and got 'OK' also with `-w` and got 'Success' but no value is changed , 
only when added `override.input.transfer.low = 150` in ups.conf and restart was working . 

```
upsrw -s input.transfer.low = 150 -u user -p pass Eaton9E2000i

upsrw -s input.eco.switchable=high-efficiency -u user -p pass Eaton9E2000i
```

But UPS was not changed to ECO Mode, still in Line mode by UPS LCD , and was no sound that switch to somewhere

masterwishx commented:
Opened issue for upsrw #2643 for not to spam here

masterwishx commented:
@arnaudquette-eaton do you think we need some think more here to set ECO mode, maybe set status also and etc...?

masterwishx commented:
From docs of Eaton 9SX : 

![image](https://github.com/user-attachments/assets/bf449d44-790a-4950-8a6e-8dab62714972)

as we already added transfer bypass/eco in #2620 , maybe we need somehow add bypass /eco tolerance monitoring befor transfer to ECO Mode?

masterwishx commented:
@jimklimov @arnaudquette-eaton can you please confirm I'm right direction and/or missing something?

jimklimov commented on drivers/usbhid-ups.h:
```diff
@@ -142,6 +143,7 @@ typedef enum {
 	BOOST,		/* SmartBoost */
 	BYPASSAUTO,	/* on automatic bypass */
 	BYPASSMAN,	/* on manual/service bypass */
+	ECOMODE,    /* High Efficiency (aka ECO Mode) */
```
I think some other code (nutconf C++? aug lens?) and configs and docs and client bindings would all need to be checked, perhaps updated, for the new `ECO` status keyword (if it is not yet there).

jimklimov replied:
Done in later commits

jimklimov commented:
Can't really help about subject matter (ECO and USB), deferring to Arnaud...

jimklimov commented:
Merged in the `LIBUSB_DEBUG` setting support, now in master...

masterwishx commented:
@desertwitch Hi, if not mistaken you have Eaton 9P Model , will be cool if you can post here debug log from When UPS in ECO Mode to see what paths its uses . 
Will be helpful i hope. If you can or/and have time for it .

Unfortunately i cant enable ECO mode on my 9E by buttons on LCD but it support it by docs and vars ...

masterwishx commented:
@arnaudquette-eaton 

`"UPS.PowerConverter.Input.[2].PresentStatus.Switchable"` is it for use with Bypass to enable disable?

arnaudquette-eaton commented:
> @arnaudquette-eaton
> 
> `"UPS.PowerConverter.Input.[2].PresentStatus.Switchable"` is it for use with Bypass to enable disable?

can't find this one

masterwishx commented:
> > @arnaudquette-eaton
> > `"UPS.PowerConverter.Input.[2].PresentStatus.Switchable"` is it for use with Bypass to enable disable?
> 
> can't find this one

```
{ "BOOL", 0, 0, "UPS.PowerConverter.Input.[2].PresentStatus.Used", NULL, NULL, 0, bypass_auto_info }, /* Automatic bypass */

also other UPS.PowerConverter.Input.[2] related to bypass in code 

```

desertwitch commented:
> @desertwitch Hi, if not mistaken you have Eaton 9P Model , will be cool if you can post here debug log from When UPS in ECO Mode to see what paths its uses . 
> 
> Will be helpful i hope. If you can or/and have time for it .
> 
> 
> 
> Unfortunately i cant enable ECO mode on my 9E by buttons on LCD but it support it by docs and vars ...

I have the 5P and I think it does not support ECO.
But will check when I am back from travels, greetings from Poland.

masterwishx commented:
> > @desertwitch Hi, if not mistaken you have Eaton 9P Model , will be cool if you can post here debug log from When UPS in ECO Mode to see what paths its uses .
> > Will be helpful i hope. If you can or/and have time for it .
> > Unfortunately i cant enable ECO mode on my 9E by buttons on LCD but it support it by docs and vars ...
> 
> I have the 5P and I think it does not support ECO. But will check when I am back from travels, greetings from Poland.

Oops, sorry for the mistake, if `5` means line interactive, then no eco mode. 
Have a good rest.

masterwishx commented:
> UPS.PowerConverter.Input.[2].PresentStatus.Switchable
Sorry my bad, i meaned `UPS.PowerConverter.Input.[2].Switchable`

masterwishx commented:
> > @arnaudquette-eaton
> > `"UPS.PowerConverter.Input.[2].PresentStatus.Switchable"` is it for use with Bypass to enable disable?
> 
> can't find this one

Sorry my bad, i meaned `UPS.PowerConverter.Input.[2].Switchable`

masterwishx commented:
@arnaudquette-eaton 
Also found: 

Path: UPS.PowerConverter.Input.[2].SwitchOffControl, Type: Feature, ReportID: 0x4f, Offset: 0, Size: 8, Value: 1
Path: UPS.PowerConverter.Input.[2].SwitchOnControl, Type: Feature, ReportID: 0x59, Offset: 0, Size: 8, Value: 0

masterwishx commented:
Also found: 
should be battery count , then must be `6` in my case for 9E2000i  but report `0`
`Path: UPS.BatterySystem.Battery.Count, Type: Feature, ReportID: 0x21, Offset: 0, Size: 8, Value: 0`

arnaudquette-eaton commented:
> @arnaudquette-eaton
> 
> in addition to : `UPS.PowerConverter.Input.[2].Switchable, Type: Feature, ReportID: 0x2d, Offset: 0, Size: 8, Value: 1`
> 
> Also found: `Path: UPS.PowerConverter.Input.[2].SwitchOffControl, Type: Feature, ReportID: 0x4f, Offset: 0, Size: 8, Value: 1`

Command to switch on inverter
0:not used, 1:switch off automatic bypass

> `Path: UPS.PowerConverter.Input.[2].SwitchOnControl, Type: Feature, ReportID: 0x59, Offset: 0, Size: 8, Value: 0`

Command to switch on bypass
0:not used, 1:switch on automatic bypass
Necessary condition to switch : 
UPS.PowerConverter.Input[2].ForcedTransferEnable equals to 1.

hope it helps, sorry but not much avail again on my side

masterwishx commented:
> > @arnaudquette-eaton
> > in addition to : `UPS.PowerConverter.Input.[2].Switchable, Type: Feature, ReportID: 0x2d, Offset: 0, Size: 8, Value: 1`
> > Also found: `Path: UPS.PowerConverter.Input.[2].SwitchOffControl, Type: Feature, ReportID: 0x4f, Offset: 0, Size: 8, Value: 1`
> 
> Command to switch on inverter 0:not used, 1:switch off automatic bypass
> 
> > `Path: UPS.PowerConverter.Input.[2].SwitchOnControl, Type: Feature, ReportID: 0x59, Offset: 0, Size: 8, Value: 0`
> 
> Command to switch on bypass 0:not used, 1:switch on automatic bypass Necessary condition to switch : UPS.PowerConverter.Input[2].ForcedTransferEnable equals to 1.
> 
> hope it helps, sorry but not much avail again on my side

Wow This will help a lot , becose not tested yet thouse values but got thinking related to bypass .

Also if i understand right by Docs Eaton UPS When switch to ECO Mode , first we need to check if Bypass (`input.bypass.voltage`) is OK and in Voltage and Freq range like `input.transfer.bypass.high/low` etc... 

if so it switches after 5min to ECO...  from Online.

If you will have time and can put eye on what i made until now (i will fix bypass commands later )  if i missed something for ECO mode like Maybe to get hardcoded in UPS status (bypass and eco) ?

masterwishx commented:
@arnaudquette-eaton 
can be that:  
`UPS.PowerConverter.Input.[5].Switchable = eco.switchable "ECO switch ability"`
`UPS.PowerConverter.Input.[2].Switchable = bypass.switchable "Bypass switch ability"`

like :  `VARDESC outlet.switchable "Outlet switch ability"`

The Eco have other switch maybe like bypass but with [5] like you wrote befor ?

masterwishx commented:
@arnaudquette-eaton @jimklimov 
! Congratulations!
We did it
Finally enabled ECO mode by software commands in 9E.
1. Enter to bypass by rw command or cmd command.
2. Wait 5 min
3. Switched rw command of eco
4. Bypass off command by cmd or rw value
5. We in OL ECO

Just need to finnalize all stuff maybe to make func to automate it...

masterwishx commented:
![20241009_102028](https://github.com/user-attachments/assets/7aca2091-3592-4c22-9fdc-ff21ba320f6b)
![20241009_102058](https://github.com/user-attachments/assets/1cf964ab-28da-44aa-b121-240c8ff9e85e)

masterwishx commented:
> hope it helps, sorry but not much avail again on my side

Without your help , it wasnt possible....

masterwishx commented:
@arnaudquette-eaton 

do you know what is Converter mode? 

![image](https://github.com/user-attachments/assets/faebe8f1-eff7-47ef-8e1d-89d27094403e)

masterwishx commented:
just for info when in ECO Mode: 

```
battery.capacity: 7.00
battery.charge: 97
battery.charge.low: 20
battery.charge.restart: 0
battery.charger.status: charging
battery.protection: yes
battery.runtime: 2861
battery.runtime.low: 180
battery.type: PbAc
battery.voltage: 83.0
battery.voltage.nominal: 72
device.mfr: EATON
device.model: 9E2000i
device.serial: xxxxxxxxxxxxxxreducted
device.type: ups
device.usb.version: 01.29
driver.debug: 0
driver.flag.allow_killpower: 0
driver.name: usbhid-ups
driver.parameter.pollfreq: 30
driver.parameter.pollinterval: 2
driver.parameter.port: auto
driver.parameter.synchronous: auto
driver.state: quiet
driver.version: 2.8.2.1
driver.version.data: MGE HID 1.49
driver.version.internal: 0.57
driver.version.usb: libusb-1.0.26 (API: 0x1000108)
input.bypass.frequency: 50.0
input.bypass.switch.off: not used
input.bypass.switch.on: not used
input.bypass.voltage: 236.0
input.eco.switchable: ecomode
input.frequency: 50.0
input.frequency.nominal: 50
input.transfer.bypass.high: 264
input.transfer.bypass.low: 184
input.transfer.eco.high: 242
input.transfer.eco.low: 218
input.transfer.forced: enabled
input.transfer.frequency.bypass.range: 10
input.transfer.frequency.eco.range: 5
input.transfer.high: 300
input.transfer.hysteresis: 10
input.transfer.low: 100
input.voltage: 235.0
input.voltage.nominal: 236
outlet.1.designator: AC OUTPUT
outlet.1.protect.status: protected
outlet.1.status: on
outlet.desc: Main Outlet
outlet.id: 0
outlet.switchable: no
output.current: 1.30
output.frequency: 50.0
output.frequency.nominal: 50
output.voltage: 235.0
output.voltage.nominal: 230
ups.alarm: HE/ECO mode!
ups.beeper.status: enabled
ups.delay.shutdown: 20
ups.delay.start: 30
ups.firmware: 00.02.000
ups.load: 16
ups.load.high: 105
ups.mfr: EATON
ups.model: 9E2000i
ups.power: 309
ups.power.nominal: 2000
ups.productid: ffff
ups.realpower: 271
ups.realpower.nominal: 1600
ups.serial:  xxxxxxxxxxxxxxreducted
ups.shutdown: enabled
ups.start.auto: yes
ups.start.battery: yes
ups.start.reboot: yes
ups.status: ALARM OL ECO
ups.temperature: 24.9
ups.test.interval: 604800
ups.test.result: Done and passed
ups.timer.shutdown: -1
ups.timer.start: -1
ups.type: online
ups.vendorid: 0463

```

jimklimov commented:
Looks great, thanks!

I hope to check later if my "Eaton Ellipse ECO 650" also supports any of this.

masterwishx commented:
> @arnaudquette-eaton
> 
> do you know what is Converter mode?
> 
> ![image](https://private-user-images.githubusercontent.com/28630321/374869317-faebe8f1-eff7-47ef-8e1d-89d27094403e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjg0NzQxNzYsIm5iZiI6MTcyODQ3Mzg3NiwicGF0aCI6Ii8yODYzMDMyMS8zNzQ4NjkzMTctZmFlYmU4ZjEtZWZmNy00N2VmLThlMWQtODlkMjcwOTQ0MDNlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMDklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDA5VDExMzc1NlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTFiNzNhY2I2YWIwMGM5NWFjNDgwOTQ0NjI5ZGU1NjdiYWUwZjdkNzU4YjUxNDUzNzI0YmUxN2RkYjViNDkyYWImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.JyLgOeMYj2HONks3-Q3CQWL_ekbQqnv9JV0D4akpplU)

Maybe its Advanced ECO mode when converter is enabled ?

masterwishx commented:
Unfortunately i can get back to online or bypass by commands or rw values , hovewer i see success of commands , maybe related to issue https://github.com/networkupstools/nut/issues/2643 .

masterwishx commented:
> UPS.PowerConverter.Input.[2].Switchable = bypass.switchable "Bypass switch ability"

@arnaudquette-eaton can it be like:
0=normal
1=qualify bypass
Etc...?

masterwishx commented:
@arnaudquette-eaton @jimklimov

As we added in previous pr bypass and eco voltage range and frequency.

1. Do we need add function to check voltage range for ECO (input.transfer.high/low) for switch back to online or ups will do it automatically?
2. We need to add also for frequency?
3. We need to make function to switch to eco if bypass is in bypass range when online or switch to bypass and check it then after 5 min switch to eco but as I have randomly some commands work or not work not shure how to test it :(

masterwishx commented:
@jimklimov 

```
upsrw -u admin -p adminpass -s input.bypass.switch.off=bypassOff -w Eaton9E2000i
Unexpected response from upsd: ERR TOO-LONG
```
upsrw -u admin -p adminpass -s input.bypass.switch.on=bypassOn -w Eaton9E2000i worked fine for me

jimklimov commented:
I've added "ECO" support to other parts of the equation (clients, notably `upsmon` reporting and docs and configs, `NEWS.adoc` etc.) and fixed spellchecker complaints.

@masterwishx : Please don't force-push over these changes :)

jimklimov commented:
Also, the changes on my side were "theoretical" so far; would be great if you can confirm that the ECO status changes are now reported by `upsmon` and are seen in UI clients (CGI, Python).

CC @desertwitch : you may also be interested in this, do I remember right?

desertwitch commented:
> Also, the changes on my side were "theoretical" so far; would be great if you can confirm that the ECO status changes are now reported by `upsmon` and are seen in UI clients (CGI, Python).
> 
> CC @desertwitch : you may also be interested in this, do I remember right?

My UPS doesn't support ECO, but will double-check logic/code-wise if that helps. 😉

masterwishx commented:
> Also, the changes on my side were "theoretical" so far; would be great if you can confirm that the ECO status changes are now reported by `upsmon` and are seen in UI clients (CGI, Python).
> 
> CC @desertwitch : you may also be interested in this, do I remember right?

I can see in plugin the status

masterwishx commented:
![image](https://github.com/user-attachments/assets/285ae7d0-529d-4dcc-9ebf-9d608e1296b8)

desertwitch commented:
Looi

> ![image](https://private-user-images.githubusercontent.com/28630321/375739515-285ae7d0-529d-4dcc-9ebf-9d608e1296b8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjg2NDcwMzIsIm5iZiI6MTcyODY0NjczMiwicGF0aCI6Ii8yODYzMDMyMS8zNzU3Mzk1MTUtMjg1YWU3ZDAtNTI5ZC00ZGNjLTllYmYtOWQ2MDhlMTI5NmI4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDExVDExMzg1MlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTIzZjIzZjc4NzliZjM1YzRjNTY5MzY4NWQyZGE0ZDY3NzBiZjE3Yzg0NmI1ZWRiMTFkNWZjMjU0MGM0OTdjNjcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.eD2nEawN2OFvqxIg3MmnpaHMNbnJ6QxF8x7-YJKG9Ng)

What's the with `ALARM` status, is this something your UPS does in ECO mode or unrelated (other) alarm?

masterwishx commented:
> Cool, what's the with `ALARM` status though? Is this something your UPS does in ECO mode or an unrelated (other) alarm?

its by ups code i can remove it in code for not use with alarm

desertwitch commented:
> > Cool, what's the with `ALARM` status though? Is this something your UPS does in ECO mode or an unrelated (other) alarm?
> 
> its by nut code i can remove alarm in code for not use in eco

I think this is something we should actively suppress now that we're handling `ECO` mode as a status, @jimklimov ?

desertwitch commented on drivers/usbhid-ups.c:
```diff
@@ -2086,6 +2093,9 @@ static void ups_alarm_set(void)
 	if (ups_status & STATUS(BYPASSMAN)) {
 		alarm_set("Manual bypass mode!");
 	}
+	if (ups_status & STATUS(ECOMODE)) {
+		alarm_set("ECO(HE) mode!");
+	}
```
I think this is something we should drop since we're handling `ECO` mode as a status? Reason being users might choose to run their UPS in ECO mode more frequently and this would raise unnecessary concerns (alarms, ha 😆 ) and/or make them "blind" to actual alarms worth being concerned about as a side-effect. 😉

jimklimov replied:
Makes sense, eventually to revise in the other driver(s?) that raise an ECO alarm...

masterwishx commented:
> > Also, the changes on my side were "theoretical" so far; would be great if you can confirm that the ECO status changes are now reported by `upsmon` and are seen in UI clients (CGI, Python).
> > CC @desertwitch : you may also be interested in this, do I remember right?
> 
> My UPS doesn't support ECO, but will double-check logic/code-wise if that helps. 😉

im trying to add function to check bypass if you can see have error here :

masterwishx commented:
So we need to check if bypass is in limits befor enter to ECO , also check eco limits befor rellaly enter to eco .  

the only thing im not shure if we can check bypass from online or we need to enter bypass and check it , after 5 min of bypass is ok we can get to eco

jimklimov commented:
FWIW, on my Eaton Ellipse ECO 650 I see few changes in `upsc` dumps:

````
 driver.parameter.vendor: EATON
 driver.parameter.vendorid: 0463
 driver.state: quiet
-driver.version: 2.8.2.1064-1064-g501bbdc62
+driver.version: 2.8.2.1085.57-1142-gf76af4d71
 driver.version.data: MGE HID 1.49
 driver.version.internal: 0.57
 driver.version.usb: libusb-1.0.24 (API: 0x1000108)
 input.transfer.high: 264
 input.transfer.low: 161
 outlet.1.desc: PowerShare Outlet 1
+outlet.1.ecocontrol: The outlet is ECO controlled
 outlet.1.id: 2
 outlet.1.status: on
 outlet.1.switchable: yes
 outlet.2.desc: PowerShare Outlet 2
+outlet.2.ecocontrol: The outlet is not ECO controlled
 outlet.2.id: 3
 outlet.2.status: on
 outlet.2.switchable: yes
````

Did not try any active commands or settings, a server is fed off this UPS now.

masterwishx commented:
> FWIW, on my Eaton Ellipse ECO 650 I see few changes in `upsc` dumps:
> 
> ```
>  driver.parameter.vendor: EATON
>  driver.parameter.vendorid: 0463
>  driver.state: quiet
> -driver.version: 2.8.2.1064-1064-g501bbdc62
> +driver.version: 2.8.2.1085.57-1142-gf76af4d71
>  driver.version.data: MGE HID 1.49
>  driver.version.internal: 0.57
>  driver.version.usb: libusb-1.0.24 (API: 0x1000108)
>  input.transfer.high: 264
>  input.transfer.low: 161
>  outlet.1.desc: PowerShare Outlet 1
> +outlet.1.ecocontrol: The outlet is ECO controlled
>  outlet.1.id: 2
>  outlet.1.status: on
>  outlet.1.switchable: yes
>  outlet.2.desc: PowerShare Outlet 2
> +outlet.2.ecocontrol: The outlet is not ECO controlled
>  outlet.2.id: 3
>  outlet.2.status: on
>  outlet.2.switchable: yes
> ```
> 
> Did not try any active commands or settings, a server is fed off this UPS now.

do you think we need disable alarm for eco mode ?

jimklimov commented:
> ```
> mge-hid.c:735:29: error: too few arguments to function ‘strtod’
>   735 |     double bypass_voltage = strtod(bypass_voltage_str);
>       |                             ^~~~~~
> In file included from ../include/common.h:58,
>                  from main.h:4,
>                  from mge-hid.c:36:
> /usr/include/stdlib.h:118:15: note: declared here
>   118 | extern double strtod (const char *__restrict __nptr,
>       |               ^~~~~~
> ````
Consider using our `str_to_double()` (from `common/str.c`)

masterwishx commented:
> FWIW, on my Eaton Ellipse ECO 650 I see few changes in `upsc` dumps:
> 
> ```
>  driver.parameter.vendor: EATON
>  driver.parameter.vendorid: 0463
>  driver.state: quiet
> -driver.version: 2.8.2.1064-1064-g501bbdc62
> +driver.version: 2.8.2.1085.57-1142-gf76af4d71
>  driver.version.data: MGE HID 1.49
>  driver.version.internal: 0.57
>  driver.version.usb: libusb-1.0.24 (API: 0x1000108)
>  input.transfer.high: 264
>  input.transfer.low: 161
>  outlet.1.desc: PowerShare Outlet 1
> +outlet.1.ecocontrol: The outlet is ECO controlled
>  outlet.1.id: 2
>  outlet.1.status: on
>  outlet.1.switchable: yes
>  outlet.2.desc: PowerShare Outlet 2
> +outlet.2.ecocontrol: The outlet is not ECO controlled
>  outlet.2.id: 3
>  outlet.2.status: on
>  outlet.2.switchable: yes
> ```
> 
> Did not try any active commands or settings, a server is fed off this UPS now.

It seems your ups is not online ups, there for you can't see new values for ECO and bypass.
As ECO mode take voltage from input bypass to output without invertor that using when online.
The ECO control on your ups is something different like economic maybe that cut off power from outlet for energy saving

masterwishx commented:
@arnaudquette-eaton any info for `Path: UPS.BatterySystem.Charger.ChargerType, Type: Feature, ReportID: 0x26, Offset: 0, Size: 8, Value: 5`?

masterwishx commented:
@jimklimov what is error about ?  

```
upsrw -u admin -p adminpass -s input.bypass.switch.off=bypassOff -w Eaton9E2000i
Unexpected response from upsd: ERR TOO-LONG
```

masterwishx commented:
> @jimklimov what is error about ?
> 
> ```
> upsrw -u admin -p adminpass -s input.bypass.switch.off=bypassOff -w Eaton9E2000i
> Unexpected response from upsd: ERR TOO-LONG
> ```

becouse changed `ecomode` to `ecoModeOn` so can throw same error

masterwishx commented:
> > FWIW, on my Eaton Ellipse ECO 650 I see few changes in `upsc` dumps:
> > ```
> >  driver.parameter.vendor: EATON
> >  driver.parameter.vendorid: 0463
> >  driver.state: quiet
> > -driver.version: 2.8.2.1064-1064-g501bbdc62
> > +driver.version: 2.8.2.1085.57-1142-gf76af4d71
> >  driver.version.data: MGE HID 1.49
> >  driver.version.internal: 0.57
> >  driver.version.usb: libusb-1.0.24 (API: 0x1000108)
> >  input.transfer.high: 264
> >  input.transfer.low: 161
> >  outlet.1.desc: PowerShare Outlet 1
> > +outlet.1.ecocontrol: The outlet is ECO controlled
> >  outlet.1.id: 2
> >  outlet.1.status: on
> >  outlet.1.switchable: yes
> >  outlet.2.desc: PowerShare Outlet 2
> > +outlet.2.ecocontrol: The outlet is not ECO controlled
> >  outlet.2.id: 3
> >  outlet.2.status: on
> >  outlet.2.switchable: yes
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Did not try any active commands or settings, a server is fed off this UPS now.
> 
> It seems your ups is not online ups, there for you can't see new values for ECO and bypass. As ECO mode take voltage from input bypass to output without invertor that using when online. The ECO control on your ups is something different like economic maybe that cut off power from outlet for energy saving

ECO Control = With an efficient electrical design and the EcoControl function
(USB models), which automatically disables peripherals when
the master device is turned off . from Eaton docs 
( the master outlet used to automatically power off the slave outlets. )

masterwishx commented:
@jimklimov please check if can be merged

jimklimov commented on drivers/mge-hid.c:
```diff
@@ -753,13 +753,13 @@ static const char *eaton_input_eco_mode_check_range(double value)
         (bypass_voltage >= eco_low && bypass_voltage <= eco_high) &&
         (bypass_frequency >= lower_frequency_limit && bypass_frequency <= upper_frequency_limit)) {
-        return "ecoModeOn"; /* Enter Eco mode */
+        return "ecoMode"; /* Enter Eco mode */
```
Why not just `ECO` for similarity with nearby `ESS`?

jimklimov commented on drivers/mge-hid.c:
```diff
@@ -832,28 +832,28 @@ static const char *eaton_input_bypass_check_range(double value)
         (bypass_voltage >= bypass_low && bypass_voltage <= bypass_high) &&
 		(bypass_frequency >= lower_frequency_limit && bypass_frequency <= upper_frequency_limit)) {
-        return "bypassOn"; /* Enter bypass mode */
+        return "On"; /* Enter bypass mode */
```
Most other such statuses deal with lower-cased `on`/`off`, Can you please change this one to match? - least-surprise principle for consumers :)

desertwitch commented on clients/status.h:
```diff
@@ -45,6 +45,7 @@ static struct {
 	{ "BOOST",	"VOLTAGE BOOST",	1	},
 	{ "CAL",	"CALIBRATION",		1	},
 	{ "BYPASS",	"BYPASS",		2	},
+	{ "ECO",	"ECO",			2	},
```
Severity 2 seems a bit excessive here (putting `ECO` in the same category as `LB`) - Severity 1 instead?

arnaudquette-eaton commented:
> @arnaudquette-eaton @jimklimov
> 
> as we added `UPS.BatterySystem.Charger.Status` here it seems we have other values from `UPS.BatterySystem.Charger.Mode` by [#2637 (comment)](https://github.com/networkupstools/nut/pull/2637#discussion_r1772730590)
> 
> that i have in 9E model , so needs to add other values and fix ABM table for .Status , maybe in other pr ?
> 
> @arnaudquette-eaton any info for: `UPS.BatterySystem.Charger.ChargerType, Type: Feature, ReportID: 0x26, Offset: 0, Size: 8, Value: 5`?

Charger type
0 - None
1 - Extended (CLA)
2 - Large extension
3 - Extra large extension (XL)
4 - ABM
5 - Constant Charge (CC)

> `UPS.BatterySystem.Charger.ChargerID, Type: Feature, ReportID: 0x20, Offset: 16, Size: 8, Value: 1`?

Item of charger collection, ID of charger
Only one charger, always 1

masterwishx commented:
> > @arnaudquette-eaton @jimklimov
> > as we added `UPS.BatterySystem.Charger.Status` here it seems we have other values from `UPS.BatterySystem.Charger.Mode` by [#2637 (comment)](https://github.com/networkupstools/nut/pull/2637#discussion_r1772730590)
> > that i have in 9E model , so needs to add other values and fix ABM table for .Status , maybe in other pr ?
> > @arnaudquette-eaton any info for: `UPS.BatterySystem.Charger.ChargerType, Type: Feature, ReportID: 0x26, Offset: 0, Size: 8, Value: 5`?
> 
> Charger type 0 - None 1 - Extended (CLA) 2 - Large extension 3 - Extra large extension (XL) 4 - ABM 5 - Constant Charge (CC)
> 
> > `UPS.BatterySystem.Charger.ChargerID, Type: Feature, ReportID: 0x20, Offset: 16, Size: 8, Value: 1`?
> 
> Item of charger collection, ID of charger Only one charger, always 1

Thanks a lot , seems we need to add this table also to ABM Table ...

> > @arnaudquette-eaton @jimklimov
> > as we added `UPS.BatterySystem.Charger.Status` here it seems we have other values from `UPS.BatterySystem.Charger.Mode` by [#2637 (comment)](https://github.com/networkupstools/nut/pull/2637#discussion_r1772730590)
> > that i have in 9E model , so needs to add other values and fix ABM table for .Status , maybe in other pr ?
> > @arnaudquette-eaton any info for: `UPS.BatterySystem.Charger.ChargerType, Type: Feature, ReportID: 0x26, Offset: 0, Size: 8, Value: 5`?
> 
> Charger type 0 - None 1 - Extended (CLA) 2 - Large extension 3 - Extra large extension (XL) 4 - ABM 5 - Constant Charge (CC)
> 
> > `UPS.BatterySystem.Charger.ChargerID, Type: Feature, ReportID: 0x20, Offset: 16, Size: 8, Value: 1`?
> 
> Item of charger collection, ID of charger Only one charger, always 1

somethink on:
`Path: UPS.PowerConverter.Input.[2].Switchable, Type: Feature, ReportID: 0x2d, Offset: 0, Size: 8, Value: 1`
should be bypass switch ability ?

masterwishx commented:
@jimklimov Maybe we can merge this pr as it's ready for ECO, if other will take more time maybe to finish?

jimklimov commented:
Bumped #2637 and #2660 against recent master changes (`ALARM` support landed into same spots almost everywhere).
NOTE: #2637 is still a strict subset of #2660.

So CI will spend some time re-checking them all...

jimklimov commented:
This PR is not as contended as the follow-up in #2660 with deeper changes into existing driver logic, so it was a good idea to separate the two :)

I intend to merge this one as soon as CI is satisfied with the recent revision.

desertwitch commented on drivers/mge-hid.c:
```diff
@@ -1286,7 +1483,10 @@ static hid_info_t mge_hid2nut[] =
 	/* Refer to Note 1 (This point will need more clarification!)
 	{ "battery.charger.status", 0, 0, "UPS.BatterySystem.Charger.PresentStatus.Used", NULL, "%.0f", HU_FLAG_QUICK_POLL, eaton_abm_enabled_legacy_info }, */
 	/* This data is the actual ABM status information */
-	{ "battery.charger.status", 0, 0, "UPS.BatterySystem.Charger.Mode", NULL, "%.0f", HU_FLAG_QUICK_POLL, eaton_abm_status_info },
+	{ "battery.charger.mode", 0, 0, "UPS.BatterySystem.Charger.Mode", NULL, "%.0f", HU_FLAG_QUICK_POLL, eaton_abm_status_info }, /* needs both ? from https://github.com/networkupstools/nut/pull/2637#discussion_r1772730590 */
+	{ "battery.charger.status", 0, 0, "UPS.BatterySystem.Charger.Status", NULL, "%.0f", HU_FLAG_QUICK_POLL, eaton_abm_status_info },
```
These changes I'd drop from this PR as not ECO-related, and overhauled in #2660 (related to ABM)

jimklimov replied:
Aha, thanks - I thought I'd checked for only-additions, but was afraid to miss something like that. Anything else?

desertwitch replied:
The rest seems ECO-related and within scope of this PR 👍

masterwishx commented:
> Just cross checked #2637 and #2660 branches against each other. #2660 seems to introduce no more ECO-changes that would be relevant to this PR, only ABM - LGTM. Might be worth renaming #2660 PR to reflect that those are ABM improvements rather than ECO-related.

Confirm #2637 for ECO/bypass mode , #2660 added statuses for ABM , sorry again that not made a separate pr for ABM

jimklimov commented:
Well, I hope to merge #2637 soon, and #2660 can then be rebased over the updated master branch or otherwise merged and reconciled, and renamed as fits.
